flux-fixpoint/src/constraint.rs

```diff
@@ -145,6 +150,31 @@ impl<Tag> Constraint<Tag> {
             Constraint::Pred(p, _) => p.is_concrete() && !p.is_trivially_true(),
         }
     }
+
+    pub fn myhash(&self) -> u64 {
+        let mut hasher = DefaultHasher::new();
+        self.hash(&mut hasher);
+        hasher.finish()
+    }
+}
+
+impl<Tag> Hash for Constraint<Tag> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        match self {
+            Constraint::Pred(p, _) => p.hash(state),
+            Constraint::Conj(cs) => cs.iter().for_each(|c| c.hash(state)),
```

Comment:
I don't fully understand the problem but maybe we should beware of prefix collision here (https://doc.rust-lang.org/std/hash/trait.Hash.html#prefix-collisions)

Reply:
hmm I don't follow that comment either -- but perhaps doesn't matter here because of the `and` ?

Reply:
I think the problems applies because if we have the following constraints

```
c1 = Constraint::Conj("a > 0", "b > 0");
c2 = Constraint::Conj("c > 0");
c3 = Constraint::Conj("a > 0");
c2 = Constraint::Conj("b > 0", "c > 0");
```

Then the the hash for the tuples `(c1, c2)` and `(c3, c2)` will be the same no?

Reply:
Being this strict will maybe miss equivalent queries that could be normalized to the same query, but better be safe than sorry.

Reply:
@nilehmann I just updated this can you take a look and see if this is what you meant?